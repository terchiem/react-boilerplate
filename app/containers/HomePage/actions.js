import {
  GET_POSTS,
  GET_POSTS_SUCCESS,
  GET_POSTS_FAIL,
  ADD_POST_TO_LIST,
} from './constants';

export function getPosts() {
  return {
    type: GET_POSTS,
  };
}

export function getPostsSuccess(posts) {
  return {
    type: GET_POSTS_SUCCESS,
    posts,
  };
}

export function getPostsFail(error) {
  return {
    type: GET_POSTS_FAIL,
    error,
  };
}

export function addPostToList(post) {
  return {
    type: ADD_POST_TO_LIST,
    post,
  };
}
