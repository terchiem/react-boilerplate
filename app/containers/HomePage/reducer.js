import produce from 'immer';
import {
  GET_POSTS,
  GET_POSTS_SUCCESS,
  GET_POSTS_FAIL,
  ADD_POST_TO_LIST,
} from './constants';

export const INITIAL_STATE = {
  loading: false,
  posts: false,
  error: false,
};

/* eslint-disable default-case, no-param-reassign */
function postListReducer(state = INITIAL_STATE, action) {
  return produce(state, draft => {
    switch (action.type) {
      case GET_POSTS:
        draft.loading = true;
        draft.posts = false;
        break;

      case GET_POSTS_SUCCESS:
        draft.loading = false;
        draft.posts = action.posts;
        break;

      case GET_POSTS_FAIL:
        draft.loading = false;
        draft.error = action.error;
        break;

      case ADD_POST_TO_LIST:
        draft.posts.unshift(action.post);
        break;

      default:
        break;
    }
    return draft;
  });
}

export default postListReducer;
