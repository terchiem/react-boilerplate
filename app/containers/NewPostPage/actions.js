import { CHANGE_POST_INPUT, RESET_POST_INPUT } from './constants';

export function changePostInput(value) {
  return {
    type: CHANGE_POST_INPUT,
    value,
  };
}

export function resetPostInput() {
  return {
    type: RESET_POST_INPUT,
  };
}
