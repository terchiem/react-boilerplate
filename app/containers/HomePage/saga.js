/**
 * Gets a list of posts from the server
 */

import request from 'utils/request';
import { call, put, takeLatest } from 'redux-saga/effects';
import { getPostsSuccess, getPostsFail } from './actions';

import { GET_POSTS } from './constants';

/**
 * Post API request/response handler
 */

export function* getPostList() {
  const requestURL = `http://localhost:3000/api`;

  try {
    const response = yield call(request, requestURL);
    yield put(getPostsSuccess(response.posts));
  } catch (err) {
    yield put(getPostsFail(err));
  }
}

export default function* getPostsSaga() {
  yield takeLatest(GET_POSTS, getPostList);
}
