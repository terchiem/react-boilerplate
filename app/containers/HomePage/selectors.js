/**
 * HomePage selectors
 */

import { createSelector } from 'reselect';
import { INITIAL_STATE } from './reducer';

export const selectHome = state => state.home || INITIAL_STATE;

export const makeSelectPosts = () =>
  createSelector(
    selectHome,
    state => state.posts,
  );

export const makeSelectLoading = () =>
  createSelector(
    selectHome,
    state => state.loading,
  );

export const makeSelectError = () =>
  createSelector(
    selectHome,
    state => state.error,
  );
