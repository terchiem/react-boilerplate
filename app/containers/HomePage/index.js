import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import injectReducer from 'utils/injectReducer';
import injectSaga from 'utils/injectSaga';

import PostList from 'components/PostList';
import { getPosts } from './actions';
import {
  makeSelectLoading,
  makeSelectPosts,
  makeSelectError,
} from './selectors';
import reducer from './reducer';
import saga from './saga';

/*
 * HomePage
 *
 * This is the first thing users see of our App, at the '/' route
 *
 * Redux state:
 *  home: {
 *    loading -> bool indicating loading state,
 *    error -> error messages,
 *    posts -> an array of posts from the server
 *  }
 *
 */

function HomePage({ loading, posts, error, dispatchGetPosts }) {
  // Dispatch GET_POSTS on component mount
  useEffect(() => {
    // Fetch posts only if not present in store
    if (posts === false) {
      dispatchGetPosts();
    }
  }, []);

  return (
    <section>
      <h2>All Posts</h2>
      <PostList loading={loading} posts={posts} error={error} />
    </section>
  );
}

HomePage.propTypes = {
  loading: PropTypes.bool,
  posts: PropTypes.oneOfType([PropTypes.bool, PropTypes.array]),
  error: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
  dispatchGetPosts: PropTypes.func,
};

const mapStateToProps = createStructuredSelector({
  loading: makeSelectLoading(),
  posts: makeSelectPosts(),
  error: makeSelectError(),
});

export const mapDispatchToProps = dispatch => ({
  dispatchGetPosts: () => dispatch(getPosts()),
});

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);
const withReducer = injectReducer({ key: 'home', reducer });
const withSaga = injectSaga({ key: 'home', saga });

export default compose(
  withConnect,
  withReducer,
  withSaga,
)(HomePage);
