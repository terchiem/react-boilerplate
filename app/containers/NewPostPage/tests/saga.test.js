/**
 * Tests for NewPostPage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';
import { push } from 'connected-react-router';
import { ADD_POST } from 'containers/HomePage/constants';
import { addPostSuccess, addPostFail } from 'containers/HomePage/actions';
import { resetPostInput } from '../actions';
import addPostSaga, { addNewPost } from '../saga';

/* eslint-disable redux-saga/yield-effects */
describe('addNewPost Saga', () => {
  let addPostGenerator;

  beforeEach(() => {
    addPostGenerator = addNewPost();

    const selectDescriptor = addPostGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = addPostGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the addPostSuccess action if it requests the data successfully', () => {
    const response = {
      post: 'Test',
    };
    const putDescriptor = addPostGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(addPostSuccess(response.post)));
    expect(addPostGenerator.next().value).toEqual(put(resetPostInput()));
    expect(addPostGenerator.next().value).toEqual(put(push('/')));
  });

  it('should call the getPostsFail action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = addPostGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(addPostFail(response)));
  });
});

describe('addPostSaga Saga', () => {
  const saga = addPostSaga();

  it('should start task to watch for ADD_POST action', () => {
    const takeLatestDescriptor = saga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(ADD_POST, addNewPost));
  });
});
