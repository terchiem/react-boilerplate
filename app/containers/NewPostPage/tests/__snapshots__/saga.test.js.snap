// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addNewPost Saga should call the getPostsFail action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`addNewPost Saga should call the getPostsFail action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3000/api",
      Object {
        "body": "{}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`addNewPost Saga should dispatch the addPostSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`addNewPost Saga should dispatch the addPostSuccess action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3000/api",
      Object {
        "body": "{}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
